File: src/lib/manifest/entityCache.ts
import { readFile, writeFile, mkdir } from "node:fs/promises";
import path from "node:path";

import type { EntityCache } from "./types/services";
import type { RawTable, RawTableName } from "./types/services";
import type { EntityCacheMeta, EntityStores, StoreName } from "./types/stores";
import { STORE_NAMES } from "./types/stores";
import { entityStorePath, entityCacheMetaPath } from "./cachePaths";
import { EXTRACTORS } from "./extractors/index";
import { buildPerkWeaponIndex, writePerkWeaponIndex } from "./perkWeaponIndex";

// ─── Node error helper ────────────────────────────────────────────────────

function isNotFoundError(err: unknown): boolean {
  return (
    typeof err === "object" &&
    err !== null &&
    (err as NodeJS.ErrnoException).code === "ENOENT"
  );
}

// ─── FileEntityCache ──────────────────────────────────────────────────────

interface FileEntityCacheOptions {
  /** The manifest version whose stores to read, or null if none built yet. */
  version: string | null;
  /** Underlying loader that reads a raw table from disk for a given version. */
  loadRawTable: (table: RawTableName) => Promise<RawTable>;
}

export class FileEntityCache implements EntityCache {
  private version: string | null;
  private readonly loadRawTable: (table: RawTableName) => Promise<RawTable>;
  private readonly storeCache = new Map<StoreName, unknown>();

  constructor(opts: FileEntityCacheOptions) {
    this.version = opts.version;
    this.loadRawTable = opts.loadRawTable;
  }

  async getMeta(): Promise<EntityCacheMeta | null> {
    if (this.version == null) return null;
    const filePath = entityCacheMetaPath(this.version);
    try {
      const text = await readFile(filePath, "utf8");
      return JSON.parse(text) as EntityCacheMeta;
    } catch (err) {
      if (isNotFoundError(err)) return null;
      throw err;
    }
  }

  async getStore<TName extends StoreName>(name: TName): Promise<EntityStores[TName]> {
    const cached = this.storeCache.get(name);
    if (cached !== undefined) return cached as EntityStores[TName];

    if (this.version == null) {
      throw new Error(
        "EntityCache: no version is set — call rebuild() first or pass a version to the constructor",
      );
    }

    const filePath = entityStorePath(this.version, name);
    let text: string;
    try {
      text = await readFile(filePath, "utf8");
    } catch (err) {
      if (isNotFoundError(err)) {
        throw new Error(
          `EntityCache: store "${name}" not found at ${filePath} — run rebuild() first`,
        );
      }
      throw err;
    }

    const store = JSON.parse(text) as EntityStores[TName];
    this.storeCache.set(name, store);
    return store;
  }

  async rebuild(version: string): Promise<EntityCacheMeta> {
    const tableCache = new Map<RawTableName, RawTable>();

    const memoizedLoad = async (table: RawTableName): Promise<RawTable> => {
      const cached = tableCache.get(table);
      if (cached !== undefined) return cached;
      const raw = await this.loadRawTable(table);
      tableCache.set(table, raw);
      return raw;
    };

    const counts = Object.fromEntries(
      STORE_NAMES.map((n) => [n, 0]),
    ) as Record<StoreName, number>;

    const builtStores = {} as EntityStores;

    for (const extractor of EXTRACTORS) {
      const data = await extractor.extract(memoizedLoad);
      const filePath = entityStorePath(version, extractor.store);
      await mkdir(path.dirname(filePath), { recursive: true });
      await writeFile(filePath, JSON.stringify(data), "utf8");
      counts[extractor.store] = (data as unknown[]).length;
      (builtStores as unknown as Record<string, unknown>)[extractor.store] = data;
    }

    const perkIndex = buildPerkWeaponIndex(version, {
      weapons: builtStores.weapons ?? [],
      "exotic-weapons": builtStores["exotic-weapons"] ?? [],
      "weapon-perks": builtStores["weapon-perks"] ?? [],
    });
    await writePerkWeaponIndex(version, perkIndex);

    const meta: EntityCacheMeta = {
      manifestVersion: version,
      builtAt: new Date().toISOString(),
      counts,
    };

    const metaPath = entityCacheMetaPath(version);
    await mkdir(path.dirname(metaPath), { recursive: true });
    await writeFile(metaPath, JSON.stringify(meta), "utf8");

    this.version = version;
    this.storeCache.clear();

    return meta;
  }
}

// ─── Factory ──────────────────────────────────────────────────────────────

export function createEntityCache(opts: FileEntityCacheOptions): EntityCache {
  return new FileEntityCache(opts);
}

File: packages/manifest/lib/src/entity_cache.dart
import 'dart:convert';
import 'dart:io';

import 'package:destiny2_storage/destiny2_storage.dart';

import 'extractors/registry.dart';
import 'types/records.dart';
import 'types/services.dart';
import 'types/stores.dart';

/// File-backed entity cache under [StorageRoot] (read + MVP rebuild).
///
/// Port of product `FileEntityCache` for MVP stores only (DART-017).
class FileEntityCache {
  FileEntityCache({
    required this.storageRoot,
    String? version,
  }) : _version = version;

  final StorageRoot storageRoot;
  String? _version;
  final Map<MvpStoreName, List<Object>> _storeCache = {};

  String? get version => _version;

  Future<EntityCacheMeta?> getMeta() async {
    final v = _version;
    if (v == null) return null;
    final file = File(storageRoot.entityCacheMetaPath(v));
    if (!await file.exists()) return null;
    final text = await file.readAsString();
    return EntityCacheMeta.fromJson(
      Map<String, dynamic>.from(jsonDecode(text) as Map),
    );
  }

  Future<List<T>> getStore<T>(MvpStoreName name) async {
    final cached = _storeCache[name];
    if (cached != null) return cached.cast<T>();

    final v = _version;
    if (v == null) {
      throw EntityCacheException(
        'no version is set — call rebuild() first or pass a version',
      );
    }

    final filePath = storageRoot.entityStorePath(v, name.fileStem);
    final file = File(filePath);
    if (!await file.exists()) {
      throw EntityCacheException(
        'store "${name.fileStem}" not found at $filePath — run rebuild() first',
      );
    }

    final text = await file.readAsString();
    final list = jsonDecode(text) as List<dynamic>;
    final records = _decodeRecords(name, list);
    _storeCache[name] = records;
    return records.cast<T>();
  }

  /// Run MVP extractors and write store JSON + meta under [StorageRoot].
  Future<EntityCacheMeta> rebuild({
    required String version,
    required LoadRawTable loadRawTable,
    DateTime? builtAt,
  }) async {
    final tableCache = <String, RawTable>{};

    Future<RawTable> memoizedLoad(String table) async {
      final cached = tableCache[table];
      if (cached != null) return cached;
      final raw = await loadRawTable(table);
      tableCache[table] = raw;
      return raw;
    }

    final counts = <String, int>{};
    final builtAtIso = (builtAt ?? DateTime.now().toUtc()).toIso8601String();

    for (final extractor in mvpExtractors) {
      final data = await extractor.extract(memoizedLoad);
      final filePath =
          storageRoot.entityStorePath(version, extractor.store.fileStem);
      await File(filePath).parent.create(recursive: true);
      final encoded = encodeStoreRecords(extractor.store, data);
      await File(filePath).writeAsString(jsonEncode(encoded));
      counts[extractor.store.fileStem] = data.length;
      _storeCache[extractor.store] = data;
    }

    final meta = EntityCacheMeta(
      manifestVersion: version,
      builtAt: builtAtIso,
      counts: counts,
    );
    final metaPath = storageRoot.entityCacheMetaPath(version);
    await File(metaPath).parent.create(recursive: true);
    await File(metaPath).writeAsString(jsonEncode(meta.toJson()));

    _version = version;
    return meta;
  }

  List<Object> _decodeRecords(MvpStoreName store, List<dynamic> list) {
    switch (store) {
      case MvpStoreName.weapons:
        return list
            .map(
              (e) => WeaponRecord.fromJson(Map<String, dynamic>.from(e as Map)),
            )
            .toList();
      case MvpStoreName.exoticArmor:
        return list
            .map(
              (e) => ExoticArmorRecord.fromJson(
                Map<String, dynamic>.from(e as Map),
              ),
            )
            .toList();
      case MvpStoreName.aspects:
        return list
            .map(
              (e) => AspectRecord.fromJson(Map<String, dynamic>.from(e as Map)),
            )
            .toList();
      case MvpStoreName.fragments:
        return list
            .map(
              (e) =>
                  FragmentRecord.fromJson(Map<String, dynamic>.from(e as Map)),
            )
            .toList();
      case MvpStoreName.abilities:
        return list
            .map(
              (e) => AbilityRecord.fromJson(Map<String, dynamic>.from(e as Map)),
            )
            .toList();
      case MvpStoreName.mods:
        return list
            .map((e) => ModRecord.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList();
    }
  }
}
